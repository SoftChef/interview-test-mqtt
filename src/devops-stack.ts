import {
  RemovalPolicy,
  Stack,
  StackProps,
} from 'aws-cdk-lib';
import {
  CfnConnection,
} from 'aws-cdk-lib/aws-codestarconnections';
import {
  PolicyStatement,
} from 'aws-cdk-lib/aws-iam';
import {
  CodePipeline,
  CodePipelineSource,
  ShellStep,
} from 'aws-cdk-lib/pipelines';
import {
  Construct,
} from 'constructs';
import {
  MqttBrokerApplication,
} from './mqtt-broker-application';

export class DevopsStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const connection = new CfnConnection(this, 'GitHubConnection', {
      connectionName: 'GitHub',
      providerType: 'GitHub',
    });
    const pipeline = new CodePipeline(this, 'Pipeline', {
      codeBuildDefaults: {
        rolePolicy: [
          new PolicyStatement({
            actions: [
              'sts:AssumeRole',
            ],
            resources: [
              'arn:aws:iam::*:role/cdk-hnb659fds-lookup-role-*',
            ],
          }),
        ],
      },
      dockerEnabledForSynth: true,
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.connection(process.env.GITHUB_REPOSITORY!, process.env.GITHUB_BRANCH_NAME!, {
          connectionArn: connection.attrConnectionArn,
          triggerOnPush: true,
        }),
        commands: [
          'yarn global add aws-cdk',
          'yarn add projen',
          'npx projen',
          'npx projen build',
          'cdk synth',
        ],
        primaryOutputDirectory: 'cdk.out',
      }),
    });
    pipeline.addStage(
      new MqttBrokerApplication(this, 'MqttBrokerApplication', {
        env: {
          account: process.env.CDK_DEPLOY_ACCOUNT,
          region: process.env.CDK_DEPLOY_REGION,
        },
      }),
    );
    connection.applyRemovalPolicy(RemovalPolicy.DESTROY);
  }
}