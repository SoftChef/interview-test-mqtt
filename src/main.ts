import {
  App,
} from 'aws-cdk-lib';
import * as dotenv from 'dotenv';
import {
  DevopsStack,
} from './devops-stack';

dotenv.config();

const app = new App();
console.log({
  account: process.env.CDK_DEPLOY_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION,
});
new DevopsStack(app, 'DevopsStack', {
  env: {
    account: process.env.CDK_DEPLOY_ACCOUNT,
    region: process.env.CDK_DEPLOY_REGION,
  },
});
app.synth();