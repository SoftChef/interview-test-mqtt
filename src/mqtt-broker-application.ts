import {
  Stage,
  StageProps,
} from 'aws-cdk-lib';
import {
  Construct,
} from 'constructs';
import {
  MqttBrokerStack,
} from './mqtt-broker-stack';

export class MqttBrokerApplication extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);
    new MqttBrokerStack(this, 'MqttBrokerStack', {
      stackName: 'MqttBrokerStack',
      env: {
        account: process.env.CDK_DEPLOY_ACCOUNT,
        region: process.env.CDK_DEPLOY_REGION,
      },
      terminationProtection: false,
    });
  }
}