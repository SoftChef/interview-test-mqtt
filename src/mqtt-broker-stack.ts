import {
  CfnOutput,
  RemovalPolicy,
  Stack,
  StackProps,
} from 'aws-cdk-lib';
import {
  Instance,
  InstanceClass,
  InstanceSize,
  InstanceType,
  MachineImage,
  Peer,
  Port,
  SecurityGroup,
  UserData,
  Vpc,
} from 'aws-cdk-lib/aws-ec2';
import {
  Construct,
} from 'constructs';

export class MqttBrokerStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const vpc = Vpc.fromLookup(this, 'Vpc', {
      isDefault: true,
    });
    const securityGroup = new SecurityGroup(this, 'MqttBrokerSecurityGroup', {
      vpc: vpc,
      allowAllOutbound: true,
    });
    // securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(22), 'SSH');
    securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(1883), 'MQTT Broker');
    const installMqttBroker: UserData = UserData.forLinux();
    installMqttBroker.addCommands('sudo apt-get update -y');
    installMqttBroker.addCommands('sudo apt-get install mosquitto -y');
    installMqttBroker.addCommands('sudo systemctl start mosquitto');
    const mqttBrokerInstance = new Instance(this, 'MqttServer', {
      instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO),
      vpc: vpc,
      machineImage: MachineImage.genericLinux({
        'ap-northeast-1': 'ami-0822295a729d2a28e',
      }),
      securityGroup: securityGroup,
      userData: installMqttBroker,
      keyName: 'InterviewTestMqtt',
    });
    new CfnOutput(this, 'MQTTBrokerEndpoint', {
      value: mqttBrokerInstance.instancePublicIp,
    });
    mqttBrokerInstance.applyRemovalPolicy(RemovalPolicy.DESTROY);
  }
}