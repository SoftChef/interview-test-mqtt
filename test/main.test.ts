import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as dotenv from 'dotenv';
import { MqttBrokerStack } from '../src/mqtt-broker-stack';

dotenv.config();

test('Snapshot', () => {
  const app = new App();
  const stack = new MqttBrokerStack(app, 'test', {
    env: {
      account: process.env.CDK_DEPLOY_ACCOUNT,
      region: process.env.CDK_DEPLOY_REGION,
    },
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});