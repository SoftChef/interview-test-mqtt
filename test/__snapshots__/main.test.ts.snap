// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "MQTTBrokerEndpoint": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "MqttServer38A23B0D",
          "PublicIp",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MqttBrokerSecurityGroup4F007042": Object {
      "Properties": Object {
        "GroupDescription": "test/MqttBrokerSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "MQTT Broker",
            "FromPort": 1883,
            "IpProtocol": "tcp",
            "ToPort": 1883,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "MqttServer38A23B0D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "MqttServerInstanceRole4849547C",
      ],
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "IamInstanceProfile": Object {
          "Ref": "MqttServerInstanceProfile65D9C7C4",
        },
        "ImageId": "ami-0822295a729d2a28e",
        "InstanceType": "t2.micro",
        "KeyName": "InterviewTestMqtt",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "MqttBrokerSecurityGroup4F007042",
              "GroupId",
            ],
          },
        ],
        "SubnetId": "p-12345",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "test/MqttServer",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
sudo apt-get update -y
sudo apt-get install mosquitto -y
sudo systemctl start mosquitto",
        },
      },
      "Type": "AWS::EC2::Instance",
      "UpdateReplacePolicy": "Delete",
    },
    "MqttServerInstanceProfile65D9C7C4": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "MqttServerInstanceRole4849547C",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "MqttServerInstanceRole4849547C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "test/MqttServer",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
